### PR TITLE
Interfaces 1.7.7

### PR DESCRIPTION
**Pull request by mjburghoffer**

### CODE REVIEW

**Comment by mjburghoffer** [comment]
@scottbessler @ajmssc @bluekelp tell me what you think

**Comment by bluekelp** [comment]
LGTM, given stylistic comments

**Comment by dvdkruk** [comment]
@mjburghoffer If I'm right, does this feature make it possible to generate Java classes (POJO style) from IDL files (indirectly using Protocol objects) and also the other way around, generate IDL file from a Protocol object? Or is it only one way?

**Comment by scottbessler** [comment]
really cool.

am i missing a way to generate JUST interfaces?

**Comment by mjburghoffer** [comment]
@scottbessler There is no way to just generate interfaces, though it would be fairly trivial to add.  The goal of this project is to generate interfaces and avro objects, and then generate 2 artifacts from this result (one is just interfaces, other is avro objects that depends on interfaces)

**Comment by scottbessler** [comment]
ah, so to generate 2 artifacts one would run a script after code gen to move to sep modules?

**Comment by mjburghoffer** [comment]
@dvdkruk the goal of this is to:
1. generate interfaces from avro schema
2. make the avro objects implement said interfaces
3. make the avro objects be able to clone any java object that implements the interfaces via `newBuilder`

If you wanted to generate pojos from these interfaces, I would suggest taking a look at the [Immutables](http://immutables.github.io/) project.  Here is an example of what I did combining this fork + immutables...

### avdl
```
@namespace("com.salesforceiq.avro")
protocol foo {

    enum TestEnum {
        FOO,
        BAR,
        BAZ
    }

    @javaAnnotation("org.immutables.value.Value.Immutable")
    record TestRecord {

        string nonnullString;
        union { null, string } @javaAnnotation("javax.annotation.Nullable") nullableString;
        int nonnullInt;
        union { null, int } @javaAnnotation("javax.annotation.Nullable") nullableInt;
        TestEnum nonnullEnum;
        union { null, TestEnum } @javaAnnotation("javax.annotation.Nullable") nullableEnum;

        array<int> intList;
        array<union { null, int }> nullableIntList;
        array<union { int, string, TestEnum }> weirdUnionList;
        array<map<array<string>>> stupidArray;

    }

}
```

### unit test
```
package com.salesforceiq.items.avro;

import com.salesforceiq.avro.TestRecord;
import com.salesforceiq.avro.common.ITestRecord;
import com.salesforceiq.avro.common.ImmutableITestRecord;
import com.salesforceiq.avro.common.TestEnum;
import org.junit.Test;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;

public class TestRecordTest {

    @Test
    public void testBuild() throws Exception {
        List<Map<String, List<String>>> stupidArray = asList(
                new HashMap<String, List<String>>() {
                    {
                        put("x", asList("y", "z"));
                        put("1", asList("2", "3"));
                    }
                },
                new HashMap<String, List<String>>() {
                    {
                        put("matt", asList("b", "m", "n"));
                        put("ascander", singletonList("d"));
                    }
                }
        );

        ITestRecord itestRecord = ImmutableITestRecord.builder()
                .nonnullEnum(TestEnum.BAR)
                .nullableEnum(TestEnum.FOO)
                .nonnullString("lol")
                .nullableString(null)
                .nonnullInt(500)
                .nullableInt(-5)
                .addIntList(-5, -55, -555, -5555, -55555)
                .nullableIntList(asList(1, 2, 3, 4, 5, null, 7))
                .weirdUnionList(asList(1, "hehe", 3, "woah", TestEnum.BAZ))
                .stupidArray(stupidArray)
                .build();


        TestRecord actual = TestRecord.newBuilder(itestRecord).build();

        TestRecord expected = TestRecord.newBuilder()
                .setNonnullEnum(TestEnum.BAR)
                .setNullableEnum(TestEnum.FOO)
                .setNonnullString("lol")
                .setNullableString(null)
                .setNonnullInt(500)
                .setNullableInt(-5)
                .setIntList(asList(-5, -55, -555, -5555, -55555))
                .setNullableIntList(asList(1, 2, 3, 4, 5, null, 7))
                .setWeirdUnionList(asList(1, "hehe", 3, "woah", TestEnum.BAZ))
                .setStupidArray(stupidArray)
                .build();

        assertEquals(expected, actual);

        stupidArray.get(0).put("a", asList("b", "c"));
        stupidArray.get(1).remove("matt");

        assertNotEquals(stupidArray, actual.getStupidArray());
    }

}
```

I create an avro object the normal way, and the interfaces way, then compare the results, everything works nicely.

**Comment by mjburghoffer** [comment]
@scottbessler I was thinking of using 2 profiles in 1 pom to handle the 2 artifacts for us.  See: http://blog.sonatype.com/2010/01/how-to-create-two-jars-from-one-project-and-why-you-shouldnt/#.VkJNamSrSRc

File: lang/java/compiler/src/main/java/org/apache/avro/compiler/specific/SpecificCompiler.java [hunk 1]
```diff
@@ -712,7 +719,31 @@ else if (nextCharToUpper) {
 
     return methodBuilder.toString();
   }
-  
+
+  public static boolean isSchemaSimpleNullable(Schema schema) {
```
Comment (scottbessler) [inline lang/java/compiler/src/main/java/org/apache/avro/compiler/specific/SpecificCompiler.java]: add a javadoc perhaps explaining what "SimpleNullable" means

File: lang/java/compiler/src/main/java/org/apache/avro/compiler/specific/SpecificCompiler.java [hunk 2]
```diff
@@ -368,35 +371,39 @@ static String makePath(String name, String space) {
     }
   }
 
-  OutputFile compile(Schema schema) {
+  List<OutputFile> compile(Schema schema) {
```
Comment (scottbessler) [inline lang/java/compiler/src/main/java/org/apache/avro/compiler/specific/SpecificCompiler.java]: :+1: for splitting these out :)

**Comment by scottbessler** [comment]
> @scottbessler I was thinking of using 2 profiles in 1 pom to handle the 2 artifacts for us. See: http://blog.sonatype.com/2010/01/how-to-create-two-jars-from-one-project-and-why-you-shouldnt/#.VkJNamSrSRc

nice. and this all looks great to me. i didnt really look into the vm templates, as I dont have much to add there but the testing you have seems fairly clear.  :heart:

**Pull request merged**